File: android/app/src/main/kotlin/com/stockbook/app/design/DropdownField.kt
package com.stockbook.app.design

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp

/**
 * "Pick one of these": a leading mark, the current choice, and a caret.
 *
 * The list is the platform's `DropdownMenu` — it already handles the scrolling,
 * the dismissal and the outside tap correctly on a phone held one-handed behind
 * a counter. Only the label is ours.
 */
@Composable
fun <T> DropdownField(
    options: List<T>,
    selected: T,
    onSelect: (T) -> Unit,
    title: (T) -> String,
    modifier: Modifier = Modifier,
    label: String? = null,
    /**
     * The short recognisable stamp on the left — a currency symbol, a language
     * code. Null where there is no useful stamp, as with a customer name.
     */
    mark: ((T) -> String)? = null,
    rowTitle: (T) -> String = title
) {
    var open by remember { mutableStateOf(false) }

    Column(modifier = modifier) {
        if (label != null) {
            Text(
                label,
                style = NocturneType.fieldLabel,
                color = Nocturne.neutral500,
                modifier = Modifier.padding(bottom = 5.dp)
            )
        }

        Box {
            Row(
                verticalAlignment = Alignment.CenterVertically,
                modifier = Modifier
                    .fillMaxWidth()
                    .height(Metrics.inputHeight)
                    .clip(RoundedCornerShape(Metrics.controlRadius))
                    .background(Nocturne.bg)
                    .hairline(radius = Metrics.controlRadius)
                    .clickable { open = true }
                    .padding(horizontal = 10.dp)
            ) {
                if (mark != null) {
                    Text(
                        mark(selected),
                        style = NocturneType.inter(15.0, FontWeight.Medium),
                        color = Nocturne.accent,
                        modifier = Modifier.widthIn(min = 34.dp)
                    )
                    Spacer(Modifier.width(9.dp))
                }
                Text(
                    title(selected),
                    style = NocturneType.inter(14.0),
                    color = Nocturne.text,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                    modifier = Modifier.weight(1f)
                )
                Spacer(Modifier.width(6.dp))
                Glyph(Icon.chooseFromList, size = 13.dp, tint = Nocturne.neutral500)
            }

            DropdownMenu(
                expanded = open,
                onDismissRequest = { open = false },
                modifier = Modifier.background(Nocturne.surface)
            ) {
                options.forEach { option ->
                    DropdownMenuItem(
                        text = {
                            Row(verticalAlignment = Alignment.CenterVertically) {
                                Glyph(
                                    Icon.confirm,
                                    size = 14.dp,
                                    tint = if (option == selected) Nocturne.accent else Nocturne.surface
                                )
                                Spacer(Modifier.width(8.dp))
                                Text(
                                    rowTitle(option),
                                    style = NocturneType.inter(14.0),
                                    color = if (option == selected) Nocturne.accent else Nocturne.text
                                )
                            }
                        },
                        onClick = {
                            open = false
                            onSelect(option)
                        }
                    )
                }
            }
        }
    }
}

File: ios/Stockbook/DesignSystem/Components/DropdownField.swift
import SwiftUI

/// The shared look of a "pick one of these" row: a leading mark, the current
/// choice, and a caret, over a `Menu` holding the list.
///
/// A `Menu` wrapping a `Picker` rather than a hand-built list. The system menu
/// already does the checkmark, the scrolling and the dismissal correctly on a
/// phone held one-handed behind a counter; the label is ours, the list is the
/// platform's.
struct DropdownField<MenuContent: View>: View {
    var label: String?
    /// The short recognisable stamp on the left — a currency symbol, a language
    /// code. Sized so both dropdowns line their titles up.
    let mark: String
    let title: String
    var accessibilityName: String
    @ViewBuilder var menu: MenuContent

    var body: some View {
        VStack(alignment: .leading, spacing: 5) {
            if let label {
                Text(label).nocturneText(.fieldLabel)
            }

            Menu {
                menu
            } label: {
                HStack(spacing: 9) {
                    Text(mark)
                        .font(NocturneType.inter(15, .medium))
                        .foregroundStyle(Nocturne.accent)
                        .frame(minWidth: 34, alignment: .leading)

                    Text(title)
                        .font(NocturneType.inter(14))
                        .foregroundStyle(Nocturne.text)
                        .lineLimit(1)

                    Spacer(minLength: 6)

                    Glyph(Icon.chooseFromList, size: 13)
                        .foregroundStyle(Nocturne.neutral500)
                }
                .padding(.horizontal, 10)
                .frame(height: Metrics.inputHeight)
                .background(Nocturne.bg, in: RoundedRectangle(cornerRadius: Metrics.controlRadius, style: .continuous))
                .hairline(radius: Metrics.controlRadius)
                .contentShape(Rectangle())
            }
            .accessibilityLabel(accessibilityName)
            .accessibilityValue(title)
        }
    }
}

/// Picking the one currency the shop bills in. Used on setup step 1 and in
/// Settings, which is the whole reason it is a component and not two blocks.
struct CurrencyField: View {
    var label: String?
    @Binding var currency: Currency

    var body: some View {
        DropdownField(
            label: label,
            // The symbol as it will appear on a bill: the part the owner
            // actually recognises. The code and name are there to confirm it.
            mark: currency.symbol.trimmed,
            title: Loc.currencyName(currency),
            accessibilityName: label ?? Loc.currencySection
        ) {
            Picker("", selection: $currency) {
                ForEach(Currency.supported) { option in
                    Text(Loc.currencyRow(option)).tag(option)
                }
            }
        }
    }
}

/// Picking the interface language.
///
/// The list is written in the languages themselves, never translated into the
/// one currently showing: somebody who has landed in a language they cannot read
/// has to be able to find their way out of it by recognising the shape of their
/// own, and "ಕನ್ನಡ" does that where "Kannada" would not.
struct LanguageField: View {
    var label: String?
    @Binding var language: AppLanguage

    var body: some View {
        DropdownField(
            label: label,
            mark: language.rawValue.uppercased(),
            title: language.endonym,
            accessibilityName: label ?? Loc.languageSection
        ) {
            Picker("", selection: $language) {
                ForEach(AppLanguage.allCases) { option in
                    Text(option.endonym).tag(option)
                }
            }
        }
    }
}
